src/node_contextify.cc

```diff
@@ -929,6 +930,138 @@ class ContextifyScript : public BaseObject {
+  Local<String> code = args[0].As<String>();
+
+  // Argument 2: filename
+  CHECK(args[1]->IsString());
+  Local<String> filename = args[1].As<String>();
+
+  // Argument 3: line offset
+  CHECK(args[2]->IsNumber());
+  Local<Integer> line_offset = args[2].As<Integer>();
+
+  // Argument 4: column offset
+  CHECK(args[3]->IsNumber());
+  Local<Integer> column_offset = args[3].As<Integer>();
+
+  // Argument 5: cached data (optional)
+  Local<Uint8Array> cached_data_buf;
+  if (!args[4]->IsUndefined()) {
+    CHECK(args[4]->IsUint8Array());
+    cached_data_buf = args[4].As<Uint8Array>();
+  }
+
+  // Argument 6: produce cache data
+  CHECK(args[5]->IsBoolean());
+  bool produce_cached_data = args[5]->IsTrue();
+
+  // Argument 7: parsing context (optional)
+  Local<Context> parsing_context;
+  if (!args[6]->IsUndefined()) {
+    CHECK(args[6]->IsObject());
+    ContextifyContext* sandbox =
+      ContextifyContext::ContextFromContextifiedSandbox(
+          env, args[6].As<Object>());
+    CHECK_NOT_NULL(sandbox);
+    parsing_context = sandbox->context();
+  } else {
+    parsing_context = context;
+  }
+
+  // Argument 8: params for the function (optional)
+  Local<Array> params_buf;
+  if (!args[7]->IsUndefined()) {
+    CHECK(args[7]->IsArray());
+    params_buf = args[7].As<Array>();
+  }
+
+  // Read cache from cached data buffer
+  ScriptCompiler::CachedData* cached_data = nullptr;
+  if (!cached_data_buf.IsEmpty()) {
+    ArrayBuffer::Contents contents = cached_data_buf->Buffer()->GetContents();
+    uint8_t* data = static_cast<uint8_t*>(contents.Data());
+    cached_data = new ScriptCompiler::CachedData(
+      data + cached_data_buf->ByteOffset(), cached_data_buf->ByteLength());
+  }
+
+  ScriptOrigin origin(filename, line_offset, column_offset);
+  ScriptCompiler::Source source(code, origin, cached_data);
+  ScriptCompiler::CompileOptions options;
+  if (source.GetCachedData() == nullptr) {
+    options = ScriptCompiler::kNoCompileOptions;
+  } else {
+    options = ScriptCompiler::kConsumeCodeCache;
+  }
+
+  TryCatch try_catch(isolate);
+  Context::Scope scope(parsing_context);
+
+  // Read params from params buffer
+  size_t params_len;
+  Local<String> *params;
+  if (!params_buf.IsEmpty()) {
+    params_len = params_buf->Length();
+    params = new Local<String>[params_len];
+
+    for (int32_t n = 0; n < params_len; n++) {
+      Local<Value> val = params_buf->Get(context, n).ToLocalChecked();
+      if (val->IsString()) {
+        params[n] = val.As<String>();
+      } else {
+        ContextifyScript::DecorateErrorStack(env, try_catch);
+        try_catch.ReThrow();
+        return;
+      }
+    }
+  } else {
+    params_len = 5;
+    params = new Local<String>[5];
+
+    params[0] = env->module_parameter_exports();
+    params[1] = env->module_parameter_require();
+    params[2] = env->module_parameter_module();
+    params[3] = env->module_parameter_filename();
+    params[4] = env->module_parameter_dirname();
```

Comment:
I still think the default should be an empty array

Reply:
@devsnek (https://github.com/ryzokuken/node/pull/1#discussion_r197889196) and @addaleax I thought so at first too, but chose otherwise because I thought eternals would be faster?

Anyway, if you're still both +1 to defaulting to an empty array instead, then I think I'm okay with it.

Reply:
> I thought eternals would be faster?

I don’t think it makes much of a difference. If you have an empty default parameter list, you don’t have *any* extra native streams, that’s also a nice thing.

But more importantly, this sets the API right to be self-contained rather than something with the specific purpose of supporting the module wrapper